create taproot descriptor template

vladimirfomene opened the pull request:
### Description

This PR solves bitcoindevkit/bdk_wallet#169. This PR adds a P2TR
descriptor template and a BIP86 taproot
descriptor template. With this, users
can now create a taproot descriptor with templates.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

w0xlt reviewed (commented):
Concept ACK

rajarshimaitra commented on src/descriptor/template.rs:
```diff
@@ -680,6 +845,8 @@ mod test {
             true,
             false,
+            false,
+            Network::Regtest,
```
I was a bit confused with this at first. Why a `Mainnet` descriptor matches with a `Regtest` address? That's because the first `network` is used for setting the 2nd derivation index of `Bipxx`, and the second network is used for the address prefix.

As a result, if someone actually tries with the same `Xpriv` but builds with `build(Network::Regtest)` and derives the addresses, they will not match up with the test vector. 

So does it make sense to change the first network into `regtest` in the build call, or change the second network to `mainnet`? The first option is a much smaller changeset and can be done in a separate PR.

vladimirfomene replied:
@rajarshimaitra , you are right this is a bit confusing. I will opt for your first option. I will option a small PR for it before end of the week. I will open an issue that helps me keep track of it.

rajarshimaitra reviewed (approved):
Review ACK 242f5bd951eef624cc7def4b654b9190bb422b8f

Everything looks good to me. Just one observation.

notmandatory commented:
@vladimirfomene please rebase to pickup the new MSRV change from bitcoindevkit/bdk#842.

notmandatory commented on src/descriptor/template.rs:
```diff
@@ -407,6 +441,85 @@ impl<K: DerivableKey<Segwitv0>> DescriptorTemplate for Bip84Public<K> {
+
+/// BIP86 public template. Expands to `tr(key/{0,1}/*)`
+///
+/// This assumes that the key used has already been derived with `m/86'/0'/0'`.
```
Should be updated in #838:
```suggestion
/// This assumes that the key used has already been derived with `m/86'/0'/0'` for Mainnet or `m/86'/1'/0'` for Testnet.
```

vladimirfomene replied:
Noted

notmandatory commented:
Would be good to get this one in before merging bdk_core work.

vladimirfomene commented:
I will work on this tonight, I will be ready tomorrow

vladimirfomene committed:
Create taproot descriptor template

This PR solves #836. This PR adds a P2TR
descriptor template and a BIP86 taproot
descriptor template. With this, users
can now create a taproot descriptor with templates.

notmandatory commented:
Thanks for all your work on this but we've gotten to the point where I'd like to hold off on merging any new features and only merge critical bug fixes to the `release/0.27` branch. 

Can you rebase this after the new `bdk_core_staging` stuff in bitcoindevkit/bdk#793 is merged to `master` branch? then we can merge it in for one of the upcoming 1.0.0-alpha releases.

danielabrozzoni commented:
Hey, can you please rebase this one on master?

notmandatory commented:
I see we didn't get merged in before the big bdk 1.0 switch. It It looks like it's ready to go, can you rebase this or update it for the new `master` branch? 

I was looking for this because @reez and I want to use a TR template for the iOS example app we're doing.

notmandatory commented:
After this is merged to master I'll get it back-ported to a maintenance release.

vladimirfomene commented:
@notmandatory , this is good to go. I have created an issue bitcoindevkit/bdk_wallet#149 for fixing the issue raised by @rajarshimaitra

notmandatory reviewed (approved):
ACK e30919ba3a7941e6f54e2f95e94533c05415d831
